File: firebase/functions/lib/admin/partner_agreements/on_partner_agreements.dart
import 'dart:async';

import 'package:firebase_admin_interop/firebase_admin_interop.dart';
import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import '../../utils/infra/firestore_event_function.dart';
import '../../utils/infra/on_firestore_helper.dart';
import '../../on_firestore_function.dart';
import 'package:data_models/community/community.dart';
import 'package:data_models/admin/partner_agreement.dart';

import '../../utils/infra/firestore_utils.dart';

class OnPartnerAgreements extends OnFirestoreFunction<PartnerAgreement> {
  OnPartnerAgreements()
      : super(
          [
            AppFirestoreFunctionData(
              'PartnerAgreementOnCreate',
              FirestoreEventType.onCreate,
            ),
            AppFirestoreFunctionData(
              'PartnerAgreementOnUpdate',
              FirestoreEventType.onUpdate,
            ),
          ],
          (snapshot) {
            return PartnerAgreement.fromJson(
              firestoreUtils.fromFirestoreJson(snapshot.data.toMap()),
            ).copyWith(id: snapshot.documentID);
          },
        );

  @override
  String get documentPath => 'partner-agreements/{partnerAgreementId}';

  @override
  Future<void> onUpdate(
    Change<DocumentSnapshot> changes,
    PartnerAgreement before,
    PartnerAgreement after,
    DateTime updateTime,
    EventContext context,
  ) async {
    print('Partner Agreement (${before.id}) has been updated');

    final beforeStripeId = before.stripeConnectedAccountId;
    final afterStripeId = after.stripeConnectedAccountId;

    // Checks whether new billing account was created.
    if (beforeStripeId == null && afterStripeId != null) {
      final String? communityId = after.communityId;
      if (communityId != null) {
        await onboardingStepsHelper.updateOnboardingSteps(
          communityId,
          changes.after,
          firestoreHelper,
          OnboardingStep.createStripeAccount,
        );
      }
    }
  }

  @override
  Future<void> onCreate(
    DocumentSnapshot documentSnapshot,
    PartnerAgreement parsedData,
    DateTime updateTime,
    EventContext context,
  ) async {
    print(
      'Partner Agreement (${documentSnapshot.documentID}) has been created',
    );

    final String communityId =
        documentSnapshot.data.getString(FirestoreHelper.kCommunityId);

    await onboardingStepsHelper.updateOnboardingSteps(
      communityId,
      documentSnapshot,
      firestoreHelper,
      OnboardingStep.createStripeAccount,
    );
  }

  @override
  Future<void> onDelete(
    DocumentSnapshot documentSnapshot,
    PartnerAgreement parsedData,
    DateTime updateTime,
    EventContext context,
  ) {
    throw UnimplementedError();
  }

  @override
  Future<void> onWrite(
    Change<DocumentSnapshot> changes,
    PartnerAgreement before,
    PartnerAgreement after,
    DateTime updateTime,
    EventContext context,
  ) {
    throw UnimplementedError();
  }
}

File: firebase/functions/src/admin/partner_agreements/on_partner_agreements.ts
import * as admin from 'firebase-admin';
import { OnFirestoreFunction, AppFirestoreFunctionData, FirestoreEventType } from '../../on_firestore_function';
import { firestoreUtils } from '../../utils/infra/firestore_utils';
import { onboardingStepsHelper } from '../../utils/onboarding_steps_helper';
import { firestoreHelper } from '../../utils/infra/on_firestore_helper';
import { OnboardingStep } from '../../types';

interface PartnerAgreement {
  id?: string;
  communityId?: string;
  stripeConnectedAccountId?: string;
  stripeConnectedAccountActive?: boolean;
  allowPayments?: boolean;
  initialUserId?: string;
  takeRate?: number;
  [key: string]: unknown;
}

type DocumentSnapshot = admin.firestore.DocumentSnapshot;

export class OnPartnerAgreements extends OnFirestoreFunction<PartnerAgreement> {
  constructor() {
    super(
      [
        { functionName: 'PartnerAgreementOnCreate', firestoreEventType: FirestoreEventType.onCreate },
        { functionName: 'PartnerAgreementOnUpdate', firestoreEventType: FirestoreEventType.onUpdate },
      ],
      (snapshot: DocumentSnapshot) => {
        return {
          ...(firestoreUtils.fromFirestoreJson(snapshot.data() ?? {}) as unknown as PartnerAgreement),
          id: snapshot.id,
        };
      }
    );
  }

  get documentPath(): string {
    return 'partner-agreements/{partnerAgreementId}';
  }

  async onUpdate(
    changes: import('firebase-functions').Change<DocumentSnapshot>,
    before: PartnerAgreement,
    after: PartnerAgreement,
    _updateTime: Date,
    context: import('firebase-functions').EventContext
  ): Promise<void> {
    console.log(`Partner Agreement (${before.id}) has been updated`);

    const beforeStripeId = before.stripeConnectedAccountId;
    const afterStripeId = after.stripeConnectedAccountId;

    if (beforeStripeId == null && afterStripeId != null) {
      const communityId = after.communityId;
      if (communityId) {
        await onboardingStepsHelper.updateOnboardingSteps(
          communityId,
          changes.after,
          firestoreHelper,
          OnboardingStep.createStripeAccount
        );
      }
    }
  }

  async onCreate(
    documentSnapshot: DocumentSnapshot,
    parsedData: PartnerAgreement,
    _updateTime: Date,
    _context: import('firebase-functions').EventContext
  ): Promise<void> {
    console.log(`Partner Agreement (${documentSnapshot.id}) has been created`);
    const communityId = (documentSnapshot.data() ?? {})['communityId'] as string | undefined;
    if (communityId) {
      await onboardingStepsHelper.updateOnboardingSteps(
        communityId,
        documentSnapshot,
        firestoreHelper,
        OnboardingStep.createStripeAccount
      );
    }
  }
}
